File: packages/core/src/engine/upset.ts
import type { UnderdogPolicy } from '../config/policy';
import { minBig } from '../math/fixed';
import { ErrorCode, fail } from './errors';
import { ONE_Q4, rawMultiplierQ4 } from './underdog';

/**
 * Upset Bonus (ECONOMICS §7.3): a pool-level top-up from the protocol Upset
 * Reserve when the winning side was the whole-Arena underdog. It never
 * touches principal and can never exceed what the reserve actually holds.
 */

export interface UpsetPolicy {
  underdog: UnderdogPolicy;
  /** Max fraction of the reserve one Arena may draw, bps. */
  reserveDrawBps: number;
  /** Absolute per-Arena cap in micro-USDC. */
  upsetBonusCapUsdc: bigint;
}

export interface UpsetInputs {
  /** Winner's whole-Arena TWAB share in bps. */
  winnerTwabShareBps: bigint;
  /** Fee- and sponsor-funded pool at settlement (before bonus). */
  basePool: bigint;
  /** Current reserve balance in micro-USDC. */
  reserveBalance: bigint;
  policy: UpsetPolicy;
}

export type UpsetLimiter = 'none' | 'formula' | 'reserveDraw' | 'reserveBalance' | 'cap';

export interface UpsetResult {
  mUpsetQ4: bigint;
  /** basePool × (m − 1) before limits. */
  requested: bigint;
  bonus: bigint;
  limitedBy: UpsetLimiter;
}

export function computeUpsetBonus(inp: UpsetInputs): UpsetResult {
  if (inp.basePool < 0n || inp.reserveBalance < 0n) fail(ErrorCode.InvalidParams, 'negative input');
  if (inp.winnerTwabShareBps < 0n || inp.winnerTwabShareBps > 10_000n) {
    fail(ErrorCode.InvalidParams, `share ${inp.winnerTwabShareBps} out of range`);
  }
  // No warm-up here: the whole-Arena integral is already the smoothing.
  const mUpsetQ4 = rawMultiplierQ4(inp.winnerTwabShareBps, inp.policy.underdog);
  const requested = (inp.basePool * (mUpsetQ4 - ONE_Q4)) / ONE_Q4;
  if (requested === 0n) return { mUpsetQ4, requested, bonus: 0n, limitedBy: 'formula' };

  const drawLimit = (inp.reserveBalance * BigInt(inp.policy.reserveDrawBps)) / 10_000n;
  const limits: Array<[bigint, UpsetLimiter]> = [
    [requested, 'none'],
    [inp.reserveBalance, 'reserveBalance'],
    [drawLimit, 'reserveDraw'],
    [inp.policy.upsetBonusCapUsdc, 'cap'],
  ];
  let bonus = requested;
  let limitedBy: UpsetLimiter = 'none';
  for (const [v, why] of limits) {
    if (v < bonus) {
      bonus = v;
      limitedBy = why;
    }
  }
  return { mUpsetQ4, requested, bonus: minBig(bonus, inp.reserveBalance), limitedBy };
}

File: programs/tribe_arena/src/engine/upset.rs
//! Upset Bonus — transcription of `packages/core/src/engine/upset.ts` (ECONOMICS §7.3).

use crate::engine::math::{to_u64, ONE_Q4};
use crate::engine::underdog::raw_multiplier_q4;
use crate::errors::TribeError;
use crate::state::UnderdogPolicy;
use anchor_lang::prelude::*;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Limiter {
    None,
    Formula,
    ReserveDraw,
    ReserveBalance,
    Cap,
}

pub struct UpsetResult {
    pub m_upset_q4: u128,
    pub requested: u64,
    pub bonus: u64,
    pub limited_by: Limiter,
}

pub fn compute_upset_bonus(
    winner_twab_share_bps: u128,
    base_pool: u64,
    reserve_balance: u64,
    underdog: &UnderdogPolicy,
    reserve_draw_bps: u16,
    upset_bonus_cap_usdc: u64,
) -> Result<UpsetResult> {
    require!(winner_twab_share_bps <= 10_000, TribeError::InvalidParams);
    let m = raw_multiplier_q4(winner_twab_share_bps, underdog);
    let requested = to_u64((base_pool as u128) * (m - ONE_Q4) / ONE_Q4)?;
    if requested == 0 {
        return Ok(UpsetResult {
            m_upset_q4: m,
            requested,
            bonus: 0,
            limited_by: Limiter::Formula,
        });
    }
    let draw_limit = to_u64((reserve_balance as u128) * (reserve_draw_bps as u128) / 10_000)?;
    // precedence: reserve balance, reserve draw, absolute cap
    let mut bonus = requested;
    let mut limited_by = Limiter::None;
    for (v, why) in [
        (reserve_balance, Limiter::ReserveBalance),
        (draw_limit, Limiter::ReserveDraw),
        (upset_bonus_cap_usdc, Limiter::Cap),
    ] {
        if v < bonus {
            bonus = v;
            limited_by = why;
        }
    }
    Ok(UpsetResult {
        m_upset_q4: m,
        requested,
        bonus: bonus.min(reserve_balance),
        limited_by,
    })
}
